### PR TITLE
Improvement update error reporting

### PR DESCRIPTION
**Pull request by windwardrail**
This PR adds some significant improvements to error reporting usefulness. It reports correct line numbers and shows textual matching information. There are still some areas for improvement though, most notably with errors that arise from mismatching optional content and superfluous nodes.

### CODE REVIEW

**Comment by windwardrail** [comment]
Whoops, forgot to resolve some warnings. Will push another commit soon.

File: src/document.rs [hunk 1]
```diff
@@ -226,6 +227,24 @@ impl<'a> Document<'a> {
...
+            .ok_or(HowserError::CapabilityError)?
+            .parent()?;
+        let line_num = node.capabilities
+            .get
+            .as_ref()
+            .ok_or(HowserError::CapabilityError)?
+            .get_start_line()?;
+
+        match (parent, line_num) {
+            (Some(parent), 0) => Document::get_line_num(&parent),
```
Comment (mullr) [inline src/document.rs]: What does it mean to get 0 back from get_start_line()?

Comment (windwardrail) [reply src/document.rs]: It means that you have an inline node or some other nested node that cmark doesn't feel like reporting an actual line number for.

Comment (windwardrail) [reply src/document.rs]: So the logic is, if I don't seem to have a line number, perhaps my parent does...

Comment (mullr) [reply src/document.rs]: :cerealspit: ok, this seems like a good thing to comment about.

File: src/errors.rs [hunk 2]
```diff
@@ -268,222 +275,344 @@ impl Reportable for DocumentError {
...
-                    _ => "Content Mismatch".to_string(),
-                };
-                break;
-            }
-        }
+        let error_type = format!(
+            "{}Type Mismatch Error{}",
+            color::Fg(color::Red),
+            color::Fg(color::Reset)
+        );
```
Comment (mullr) [inline src/errors.rs]: This formatting pattern is repeated a lot, good opportunity for a little helper.

File: src/errors.rs [hunk 3]
```diff
@@ -268,222 +275,344 @@ impl Reportable for DocumentError {
...
             .as_ref()
-            .unwrap_or(&String::new())
+            .unwrap_or(&"Unknown".to_string())
             .to_string();
+        let doc_line = Document::get_line_num(doc_node)?;
+
+        let rx_prompts: Vec<String> = match_pairs
+            .iter()
+            .map(|pair| {
+                let content = pair.0.to_string();
```
[the remark this answers is not in the document]
Comment (mullr) [reply src/errors.rs]: oh, perhaps you can't because it's used directly on the next line.

File: src/errors.rs [hunk 4]
```diff
@@ -268,222 +275,344 @@ impl Reportable for DocumentError {
...
+            .collect();
+        let doc_matches: Vec<String> = match_pairs
+            .iter()
+            .map(|pair| {
+                let content = pair.1
+                    .as_ref()
+                    .unwrap_or(&String::from("<No Match>"))
+                    .to_owned();
+                match ContentMatchPair::is_match(pair) {
+                    true => format!(
```
Comment (mullr) [inline src/errors.rs]: formatty mcformatmace

File: src/errors.rs [hunk 5]
```diff
@@ -268,222 +275,344 @@ impl Reportable for DocumentError {
...
+            self.rx_line,
+            style::Reset
+        );
+        let doc_info = format!(
+            "{}{} line {}{}",
+            style::Underline,
+            self.doc_file,
+            self.doc_line,
+            style::Reset
+        );
```
Comment (mullr) [inline src/errors.rs]: I feel like I just read this code; is there any way you can take these two error structs and factor out the common bits? The code to deal with them seems more alike than different.

File: src/document.rs (hunk 1 above)
Comment (windwardrail) [reply src/document.rs]: Commented

File: src/errors.rs [hunk 6]
```diff
@@ -268,222 +275,344 @@ impl Reportable for DocumentError {
...
+
+        message += &format!("{}Prescription : {}", style::Bold, style::Reset);
+        message += &self.rx_prompts
+            .iter()
+            .flat_map(|s| s.chars())
+            .collect::<String>();
```
Comment (mullr) [inline src/errors.rs]: This is concatenating the strings together, right? This cannot be fast; surely there's a better way.

File: src/helpers/cli/mod.rs [hunk 7]
```diff
@@ -36,17 +41,22 @@ impl ShellText {
...
+            )
         })
         .enumerate()
         .map(|(i, line)| {
-            let line_num = ShellText::Dim(Box::new(ShellText::Literal(
-                format!("{}    ", i as u32 + start_line)[..4].to_string(),
-            )));
-            format!("{} {}", line_num.to_string(), line)
+            let mut line_num = format!("{}{}    ", style::Faint, i + start_line);
+            line_num.truncate(6);
```
Comment (mullr) [inline src/helpers/cli/mod.rs]: This looks suspicious to me. What are you trying to do?

File: src/errors.rs (hunk 2 above)
Comment (windwardrail) [reply src/errors.rs]: helpers added

File: src/errors.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (windwardrail) [reply src/errors.rs]: Yeah, it's a little ugly, but not as much as destructuring and then reassembling a couple lines later imo

File: src/errors.rs [hunk 8]
```diff
@@ -268,222 +275,344 @@ impl Reportable for DocumentError {
...
+                        content,
+                        color::Fg(color::Reset)
+                    ),
+                }
+            })
+            .collect();
+        let doc_matches: Vec<String> = match_pairs
+            .iter()
+            .map(|pair| {
+                let content = pair.1
```
[the remark this answers is not in the document]
Comment (windwardrail) [reply src/errors.rs]: No, it's correct. The first element corresponds to an rx prompt and the second corresponds to text from the document that is supposed to match it.

File: src/errors.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mullr) [reply src/errors.rs]: I believe you could do `|pair @ (content, _)|`

[the remark this answers is not in the document]
Comment (mullr) [reply src/errors.rs]: meh, probably not worth it at that point

File: src/helpers/cli/mod.rs (hunk 7 above)
Comment (mullr) [reply src/helpers/cli/mod.rs]: I'd expect rust to have width-sensitive format specifiers, like `printf` does.

Comment (windwardrail) [reply src/helpers/cli/mod.rs]: Updated to use format specifier

**Pull request merged**